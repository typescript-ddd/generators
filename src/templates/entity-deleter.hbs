import { DomainEventBus, EntityNotFoundError } from "@typescript-ddd/core";
import { {{pascal entityName}}Repository } from "../../infrastructure/persistence";
import { {{pascal entityName}} } from "../../domain";

/**
 * Represents {{noun entityName}} deleter.
 */
export class {{pascal entityName}}Deleter {
  /**
   * Initializes a new instance of {{noun entityName}} deleter.
   * @param {{lbrace}}{{pascal entityName}}Repository{{rbrace}} repository - The {{natural entityName}} repository.
   * @param {DomainEventBus} eventBus - The domain event bus.
   */
  constructor(
    private readonly repository: {{pascal entityName}}Repository,
    private readonly eventBus: DomainEventBus
  ) {}

  /**
   * Deletes {{noun entityName}}.
   * @param {{lbrace}}{{pascal entityName}}Id{{rbrace}} id - The {{natural entityName}} id.
   * @returns {Promise<void>}
   */
  public async delete(id: {{pascal entityName}}Id): Promise<void> {
    const entity = await this.repository.findById(id);
    if (!entity) {
      throw new EntityNotFoundError("{{pascal entityName}}", id);
    }
    entity.delete();

    await this.repository.delete(id);
    this.eventBus.publish(entity.pullEvents());
  }
}
