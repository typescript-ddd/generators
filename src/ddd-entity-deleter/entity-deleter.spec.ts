import { EntityDeleterTemplateData } from "./entity-deleter.types";
import { generateEntityDeleter } from "./entity-deleter.generator";

describe("EntityDeleter", () => {
    it("should render an entity deleter template", () => {
        // Arrange
        const data: EntityDeleterTemplateData = {
        entityName: "User"
        };
    
        // Act
        const result = generateEntityDeleter(data);
    
        // Assert
        expect(result).toMatchSnapshot();
    });
    });