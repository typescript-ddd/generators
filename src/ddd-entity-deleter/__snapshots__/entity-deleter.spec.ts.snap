// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityDeleter should render an entity deleter template 1`] = `
"import { DomainEventBus, EntityNotFoundError } from "@typescript-ddd/core";
import { UserRepository } from "../../infrastructure/persistence";
import { User } from "../../domain";

/**
 * Represents a user deleter.
 */
export class UserDeleter {
  /**
   * Initializes a new instance of a user deleter.
   * @param {UserRepository} repository - The user repository.
   * @param {DomainEventBus} eventBus - The domain event bus.
   */
  constructor(
    private readonly repository: UserRepository,
    private readonly eventBus: DomainEventBus
  ) {}

  /**
   * Deletes a user.
   * @param {UserId} id - The user id.
   * @returns {Promise<void>}
   */
  public async delete(id: UserId): Promise<void> {
    const entity = await this.repository.findById(id);
    if (!entity) {
      throw new EntityNotFoundError("User", id);
    }
    entity.delete();

    await this.repository.delete(id);
    this.eventBus.publish(entity.pullEvents());
  }
}
"
`;
